ICE: impl_trait_in_assoc_type: `expected type for .. but found Lifetime ...`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(impl_trait_in_assoc_type)]
#![crate_type="lib"]

use std::future::Future;

trait Stream {}

trait X {
    type LineStream<'a, Repr>
    where
        Self: 'a;
    type LineStreamFut<'a, Repr>
    where
        Self: 'a;
}

struct Y;

impl X for Y {
    type LineStream<T> = impl Stream;
    type LineStreamFut<'a, Repr> = impl Future<Output = Self::LineStream<'a, Repr>>;
    fn line_stream<'a, Repr>(&'a self) -> Self::LineStreamFut<'a, Repr> {}
}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (77f4f828a 2023-05-20)
binary: rustc
commit-hash: 77f4f828a2f19854fcbcdf69babe7d0ac1c92852
commit-date: 2023-05-20
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.4
```

### Error output

```
error[E0407]: method `line_stream` is not a member of trait `X`
  --> treereduce.out:22:5
   |
22 |     fn line_stream<'a, Repr>(&'a self) -> Self::LineStreamFut<'a, Repr> {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a member of trait `X`

error[E0195]: lifetime parameters or bounds on type `LineStream` do not match the trait declaration
  --> treereduce.out:20:20
   |
9  |     type LineStream<'a, Repr>
   |                    ---------- lifetimes in impl do not match this type in trait
10 |     where
11 |         Self: 'a;
   |               -- this bound might be missing in the impl
...
20 |     type LineStream<T> = impl Stream;
   |                    ^^^ lifetimes do not match type in trait
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_middle/src/ty/subst.rs:878:9: expected type for `T/#0` (T/0) but found Lifetime(ReLateBound(DebruijnIndex(1), BoundRegion { var: 0, kind: BrNamed(DefId(0:21 ~ treereduce[171b]::{impl#0}::line_stream::'a), 'a) })) when substituting, substs=[ReLateBound(DebruijnIndex(1), BoundRegion { var: 0, kind: BrNamed(DefId(0:21 ~ treereduce[171b]::{impl#0}::line_stream::'a), 'a) }), Repr]

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/compiler/rustc_errors/src/lib.rs:1651:9
stack backtrace:
   0:     0x7f1ea55698c1 - std::backtrace_rs::backtrace::libunwind::trace::hb596ab92dd1ffeb1
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f1ea55698c1 - std::backtrace_rs::backtrace::trace_unsynchronized::h6f83052e2e74814e
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f1ea55698c1 - std::sys_common::backtrace::_print_fmt::h194236fb3bd0ec03
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f1ea55698c1 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcb9b1916f3013664
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f1ea55c9dbf - core::fmt::rt::Argument::fmt::h06db2505b86cd847
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/core/src/fmt/rt.rs:138:9
   5:     0x7f1ea55c9dbf - core::fmt::write::h6e3a67ff7ee8c309
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f1ea555cbc1 - std::io::Write::write_fmt::he7dbf2095d7fa761
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/io/mod.rs:1712:15
   7:     0x7f1ea55696d5 - std::sys_common::backtrace::_print::hb8679104c9514b7f
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f1ea55696d5 - std::sys_common::backtrace::print::h780e3684add454c8
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f1ea556c347 - std::panicking::default_hook::{{closure}}::h9ab5a89be5bee36f
  10:     0x7f1ea556c134 - std::panicking::default_hook::hc5acf7e0e03a470a
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/panicking.rs:288:9
  11:     0x7f1ea86eecbb - rustc_driver_impl[69d5d36b0a095829]::install_ice_hook::{closure#0}
  12:     0x7f1ea556ca67 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4beb3e09651d2541
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/alloc/src/boxed.rs:1999:9
  13:     0x7f1ea556ca67 - std::panicking::rust_panic_with_hook::h1b1a8bf972db194d
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/panicking.rs:695:13
  14:     0x7f1ea8bf7631 - std[e8677ec1892d62a0]::panicking::begin_panic::<rustc_errors[bc4c26190069b25c]::ExplicitBug>::{closure#0}
  15:     0x7f1ea8bf1436 - std[e8677ec1892d62a0]::sys_common::backtrace::__rust_end_short_backtrace::<std[e8677ec1892d62a0]::panicking::begin_panic<rustc_errors[bc4c26190069b25c]::ExplicitBug>::{closure#0}, !>
  16:     0x7f1ea8bed456 - std[e8677ec1892d62a0]::panicking::begin_panic::<rustc_errors[bc4c26190069b25c]::ExplicitBug>
  17:     0x7f1ea8bea9f4 - <rustc_errors[bc4c26190069b25c]::HandlerInner>::bug::<alloc[fbb51881c3502246]::string::String>
  18:     0x7f1ea8bea796 - <rustc_errors[bc4c26190069b25c]::Handler>::bug::<alloc[fbb51881c3502246]::string::String>
  19:     0x7f1ea8c6e20c - rustc_middle[207a24e714ea4d12]::util::bug::opt_span_bug_fmt::<rustc_span[1c8b785102123e08]::span_encoding::Span>::{closure#0}
  20:     0x7f1ea8c6b6fa - rustc_middle[207a24e714ea4d12]::ty::context::tls::with_opt::<rustc_middle[207a24e714ea4d12]::util::bug::opt_span_bug_fmt<rustc_span[1c8b785102123e08]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f1ea8c6b6ca - rustc_middle[207a24e714ea4d12]::ty::context::tls::with_context_opt::<rustc_middle[207a24e714ea4d12]::ty::context::tls::with_opt<rustc_middle[207a24e714ea4d12]::util::bug::opt_span_bug_fmt<rustc_span[1c8b785102123e08]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f1ea6e68e7d - rustc_middle[207a24e714ea4d12]::util::bug::bug_fmt
  23:     0x7f1ea8c01ec2 - <rustc_middle[207a24e714ea4d12]::ty::subst::SubstFolder>::type_param_expected
  24:     0x7f1ea67aaa41 - <&rustc_middle[207a24e714ea4d12]::ty::list::List<rustc_middle[207a24e714ea4d12]::ty::subst::GenericArg> as rustc_type_ir[9cb919233080541a]::fold::TypeFoldable<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::try_fold_with::<rustc_middle[207a24e714ea4d12]::ty::subst::SubstFolder>
  25:     0x7f1ea67a2a2a - <rustc_middle[207a24e714ea4d12]::ty::Ty as rustc_type_ir[9cb919233080541a]::fold::TypeSuperFoldable<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::super_fold_with::<rustc_middle[207a24e714ea4d12]::ty::subst::SubstFolder>
  26:     0x7f1ea93dd6b0 - <rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector as rustc_type_ir[9cb919233080541a]::visit::TypeVisitor<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::visit_ty
  27:     0x7f1ea93f2463 - <rustc_middle[207a24e714ea4d12]::ty::Term as rustc_type_ir[9cb919233080541a]::visit::TypeVisitable<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::visit_with::<rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector>
  28:     0x7f1ea93f1e0e - <rustc_middle[207a24e714ea4d12]::ty::PredicateKind as rustc_type_ir[9cb919233080541a]::visit::TypeVisitable<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::visit_with::<rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector>
  29:     0x7f1ea93dd521 - <rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector as rustc_type_ir[9cb919233080541a]::visit::TypeVisitor<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::visit_ty
  30:     0x7f1ea93dd6bc - <rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector as rustc_type_ir[9cb919233080541a]::visit::TypeVisitor<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::visit_ty
  31:     0x7f1ea93f97f3 - <rustc_middle[207a24e714ea4d12]::ty::sty::Binder<&rustc_middle[207a24e714ea4d12]::ty::list::List<rustc_middle[207a24e714ea4d12]::ty::Ty>> as rustc_type_ir[9cb919233080541a]::visit::TypeSuperVisitable<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::super_visit_with::<rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector>
  32:     0x7f1ea93dce8a - <rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::OpaqueTypeCollector>::collect::<rustc_middle[207a24e714ea4d12]::ty::sty::FnSig>
  33:     0x7f1ea93dcffe - rustc_ty_utils[c6c31b3f05f4b983]::opaque_types::opaque_types_defined_by
  34:     0x7f1ea8ffa86f - rustc_query_impl[2598de8647e50398]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2598de8647e50398]::query_impl::opaque_types_defined_by::dynamic_query::{closure#2}::{closure#0}, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 16usize]>>
  35:     0x7f1ea8fd9db2 - <rustc_query_impl[2598de8647e50398]::query_impl::opaque_types_defined_by::dynamic_query::{closure#2} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<(rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt, rustc_span[1c8b785102123e08]::def_id::LocalDefId)>>::call_once
  36:     0x7f1ea700ff86 - rustc_query_system[584a3473f56e3e7f]::query::plumbing::try_execute_query::<rustc_query_impl[2598de8647e50398]::DynamicConfig<rustc_query_system[584a3473f56e3e7f]::query::caches::VecCache<rustc_span[1c8b785102123e08]::def_id::LocalDefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[2598de8647e50398]::plumbing::QueryCtxt, false>
  37:     0x7f1ea8fac79c - rustc_query_impl[2598de8647e50398]::query_impl::opaque_types_defined_by::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7f1ea75d51c7 - <rustc_infer[67d55827fc61bf78]::infer::InferCtxt>::opaque_type_origin
  39:     0x7f1ea7e3c591 - <rustc_middle[207a24e714ea4d12]::ty::fold::BottomUpFolder<<rustc_infer[67d55827fc61bf78]::infer::InferCtxt>::replace_opaque_types_with_inference_vars<rustc_middle[207a24e714ea4d12]::ty::Ty>::{closure#3}, <rustc_infer[67d55827fc61bf78]::infer::InferCtxt>::replace_opaque_types_with_inference_vars<rustc_middle[207a24e714ea4d12]::ty::Ty>::{closure#1}, <rustc_infer[67d55827fc61bf78]::infer::InferCtxt>::replace_opaque_types_with_inference_vars<rustc_middle[207a24e714ea4d12]::ty::Ty>::{closure#2}> as rustc_type_ir[9cb919233080541a]::fold::FallibleTypeFolder<rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt>>::try_fold_ty
  40:     0x7f1ea72251d6 - rustc_hir_typeck[bc8e5ff3546f838d]::check::check_fn
  41:     0x7f1ea720cb6a - rustc_hir_typeck[bc8e5ff3546f838d]::typeck
  42:     0x7f1ea67cdf1e - rustc_query_impl[2598de8647e50398]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2598de8647e50398]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>
  43:     0x7f1ea67cdeee - <rustc_query_impl[2598de8647e50398]::query_impl::typeck::dynamic_query::{closure#2} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<(rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt, rustc_span[1c8b785102123e08]::def_id::LocalDefId)>>::call_once
  44:     0x7f1ea6898490 - rustc_query_system[584a3473f56e3e7f]::query::plumbing::try_execute_query::<rustc_query_impl[2598de8647e50398]::DynamicConfig<rustc_query_system[584a3473f56e3e7f]::query::caches::VecCache<rustc_span[1c8b785102123e08]::def_id::LocalDefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[2598de8647e50398]::plumbing::QueryCtxt, false>
  45:     0x7f1ea80b2901 - rustc_query_impl[2598de8647e50398]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  46:     0x7f1ea882c777 - rustc_middle[207a24e714ea4d12]::query::plumbing::query_get_at::<rustc_query_system[584a3473f56e3e7f]::query::caches::VecCache<rustc_span[1c8b785102123e08]::def_id::LocalDefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>>
  47:     0x7f1ea88307f8 - <rustc_hir_analysis[425b9e45a9ba480c]::collect::type_of::find_opaque_ty_constraints_for_tait::ConstraintLocator>::check
  48:     0x7f1ea882ee31 - <rustc_hir_analysis[425b9e45a9ba480c]::collect::type_of::find_opaque_ty_constraints_for_tait::ConstraintLocator as rustc_hir[6ef0987b02b73b3]::intravisit::Visitor>::visit_nested_impl_item
  49:     0x7f1ea8826ca5 - rustc_hir[6ef0987b02b73b3]::intravisit::walk_item::<rustc_hir_analysis[425b9e45a9ba480c]::collect::type_of::find_opaque_ty_constraints_for_tait::ConstraintLocator>
  50:     0x7f1ea88305ea - rustc_hir_analysis[425b9e45a9ba480c]::collect::type_of::find_opaque_ty_constraints_for_tait
  51:     0x7f1ea7afc918 - rustc_hir_analysis[425b9e45a9ba480c]::collect::type_of::type_of
  52:     0x7f1ea69009c3 - rustc_query_impl[2598de8647e50398]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2598de8647e50398]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>
  53:     0x7f1ea6900980 - <rustc_query_impl[2598de8647e50398]::query_impl::type_of::dynamic_query::{closure#2} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<(rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt, rustc_span[1c8b785102123e08]::def_id::DefId)>>::call_once
  54:     0x7f1ea67cf9c7 - rustc_query_system[584a3473f56e3e7f]::query::plumbing::try_execute_query::<rustc_query_impl[2598de8647e50398]::DynamicConfig<rustc_query_system[584a3473f56e3e7f]::query::caches::DefaultCache<rustc_span[1c8b785102123e08]::def_id::DefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[2598de8647e50398]::plumbing::QueryCtxt, false>
  55:     0x7f1ea80a83b3 - rustc_query_impl[2598de8647e50398]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  56:     0x7f1ea68c7760 - rustc_middle[207a24e714ea4d12]::query::plumbing::query_get_at::<rustc_query_system[584a3473f56e3e7f]::query::caches::DefaultCache<rustc_span[1c8b785102123e08]::def_id::DefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 8usize]>>>
  57:     0x7f1ea68c1805 - rustc_hir_analysis[425b9e45a9ba480c]::check::check::check_mod_item_types
  58:     0x7f1ea73ba39c - rustc_query_impl[2598de8647e50398]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2598de8647e50398]::query_impl::check_mod_item_types::dynamic_query::{closure#2}::{closure#0}, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 0usize]>>
  59:     0x7f1ea73ba37e - <rustc_query_impl[2598de8647e50398]::query_impl::check_mod_item_types::dynamic_query::{closure#2} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<(rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt, rustc_span[1c8b785102123e08]::def_id::LocalDefId)>>::call_once
  60:     0x7f1ea6f2f8ef - rustc_query_system[584a3473f56e3e7f]::query::plumbing::try_execute_query::<rustc_query_impl[2598de8647e50398]::DynamicConfig<rustc_query_system[584a3473f56e3e7f]::query::caches::VecCache<rustc_span[1c8b785102123e08]::def_id::LocalDefId, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[2598de8647e50398]::plumbing::QueryCtxt, false>
  61:     0x7f1ea80bbff4 - rustc_query_impl[2598de8647e50398]::query_impl::check_mod_item_types::get_query_non_incr::__rust_end_short_backtrace
  62:     0x7f1ea7bd3857 - <rustc_middle[207a24e714ea4d12]::hir::map::Map>::for_each_module::<rustc_hir_analysis[425b9e45a9ba480c]::check_crate::{closure#6}::{closure#0}>
  63:     0x7f1ea7bd2c8d - <rustc_session[b9bbdcb4e08edc4b]::session::Session>::time::<(), rustc_hir_analysis[425b9e45a9ba480c]::check_crate::{closure#6}>
  64:     0x7f1ea7bd0bc7 - rustc_hir_analysis[425b9e45a9ba480c]::check_crate
  65:     0x7f1ea7bcb8dd - rustc_interface[bf06fa521c2e14d4]::passes::analysis
  66:     0x7f1ea7c2b55a - rustc_query_impl[2598de8647e50398]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2598de8647e50398]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 1usize]>>
  67:     0x7f1ea7c2b549 - <rustc_query_impl[2598de8647e50398]::query_impl::analysis::dynamic_query::{closure#2} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<(rustc_middle[207a24e714ea4d12]::ty::context::TyCtxt, ())>>::call_once
  68:     0x7f1ea7dfc464 - rustc_query_system[584a3473f56e3e7f]::query::plumbing::try_execute_query::<rustc_query_impl[2598de8647e50398]::DynamicConfig<rustc_query_system[584a3473f56e3e7f]::query::caches::SingleCache<rustc_middle[207a24e714ea4d12]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[2598de8647e50398]::plumbing::QueryCtxt, false>
  69:     0x7f1ea7dfc249 - rustc_query_impl[2598de8647e50398]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  70:     0x7f1ea7dc365f - <rustc_middle[207a24e714ea4d12]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[69d5d36b0a095829]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>>
  71:     0x7f1ea7983edf - <rustc_interface[bf06fa521c2e14d4]::interface::Compiler>::enter::<rustc_driver_impl[69d5d36b0a095829]::run_compiler::{closure#1}::{closure#2}, core[fd54cc11b4e2c08d]::result::Result<core[fd54cc11b4e2c08d]::option::Option<rustc_interface[bf06fa521c2e14d4]::queries::Linker>, rustc_span[1c8b785102123e08]::ErrorGuaranteed>>
  72:     0x7f1ea798172a - std[e8677ec1892d62a0]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[bf06fa521c2e14d4]::util::run_in_thread_pool_with_globals<rustc_interface[bf06fa521c2e14d4]::interface::run_compiler<core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>, rustc_driver_impl[69d5d36b0a095829]::run_compiler::{closure#1}>::{closure#0}, core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>>
  73:     0x7f1ea7fad835 - <<std[e8677ec1892d62a0]::thread::Builder>::spawn_unchecked_<rustc_interface[bf06fa521c2e14d4]::util::run_in_thread_pool_with_globals<rustc_interface[bf06fa521c2e14d4]::interface::run_compiler<core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>, rustc_driver_impl[69d5d36b0a095829]::run_compiler::{closure#1}>::{closure#0}, core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fd54cc11b4e2c08d]::result::Result<(), rustc_span[1c8b785102123e08]::ErrorGuaranteed>>::{closure#1} as core[fd54cc11b4e2c08d]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  74:     0x7f1ea5576f15 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hbb95789317ed1f9b
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/alloc/src/boxed.rs:1985:9
  75:     0x7f1ea5576f15 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h52fc295245ab71ea
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/alloc/src/boxed.rs:1985:9
  76:     0x7f1ea5576f15 - std::sys::unix::thread::Thread::new::thread_start::hecdeb81401fa1c6c
                               at /rustc/77f4f828a2f19854fcbcdf69babe7d0ac1c92852/library/std/src/sys/unix/thread.rs:108:17
  77:     0x7f1ea5313bb5 - <unknown>
  78:     0x7f1ea5395d90 - <unknown>
  79:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (77f4f828a 2023-05-20) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [opaque_types_defined_by] computing the opaque types defined by `<impl at treereduce.out:19:1: 19:13>::line_stream`
#1 [typeck] type-checking `<impl at treereduce.out:19:1: 19:13>::line_stream`
#2 [type_of] computing type of `<impl at treereduce.out:19:1: 19:13>::LineStream::{opaque#0}`
#3 [check_mod_item_types] checking item types in top-level module
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0195, E0407.
For more information about an error, try `rustc --explain E0195`.
```

</p>
</details>

--- a/compiler/rustc_trait_selection/src/traits/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/mod.rs
@@ -65,12 +65,12 @@ pub use self::specialize::{
 pub use self::structural_match::search_for_structural_match_violation;
 pub use self::structural_normalize::StructurallyNormalizeExt;
 pub use self::util::elaborate;
+pub use self::util::{
+    check_substs_compatible, supertrait_def_ids, supertraits, transitive_bounds,
+    transitive_bounds_that_define_assoc_item, SupertraitDefIds,
+};
 pub use self::util::{expand_trait_aliases, TraitAliasExpander};
 pub use self::util::{get_vtable_index_of_object_method, impl_item_is_final, upcast_choices};
-pub use self::util::{
-    supertrait_def_ids, supertraits, transitive_bounds, transitive_bounds_that_define_assoc_item,
-    SupertraitDefIds,
-};
 
 pub use self::chalk_fulfill::FulfillmentContext as ChalkFulfillmentContext;
 

--- a/compiler/rustc_trait_selection/src/traits/project.rs
+++ b/compiler/rustc_trait_selection/src/traits/project.rs
@@ -1,5 +1,6 @@
 //! Code for projecting associated types out of trait references.
 
+use super::check_substs_compatible;
 use super::specialization_graph;
 use super::translate_substs;
 use super::util;
@@ -2376,47 +2377,6 @@ fn confirm_impl_candidate<'cx, 'tcx>(
         assoc_ty_own_obligations(selcx, obligation, &mut nested);
         Progress { term: term.subst(tcx, substs), obligations: nested }
     }
-}
-
-// Verify that the trait item and its implementation have compatible substs lists
-fn check_substs_compatible<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    assoc_item: ty::AssocItem,
-    substs: ty::SubstsRef<'tcx>,
-) -> bool {
-    fn check_substs_compatible_inner<'tcx>(
-        tcx: TyCtxt<'tcx>,
-        generics: &'tcx ty::Generics,
-        args: &'tcx [ty::GenericArg<'tcx>],
-    ) -> bool {
-        if generics.count() != args.len() {
-            return false;
-        }
-
-        let (parent_args, own_args) = args.split_at(generics.parent_count);
-
-        if let Some(parent) = generics.parent
-            && let parent_generics = tcx.generics_of(parent)
-            && !check_substs_compatible_inner(tcx, parent_generics, parent_args) {
-            return false;
-        }
-
-        for (param, arg) in std::iter::zip(&generics.params, own_args) {
-            match (&param.kind, arg.unpack()) {
-                (ty::GenericParamDefKind::Type { .. }, ty::GenericArgKind::Type(_))
-                | (ty::GenericParamDefKind::Lifetime, ty::GenericArgKind::Lifetime(_))
-                | (ty::GenericParamDefKind::Const { .. }, ty::GenericArgKind::Const(_)) => {}
-                _ => return false,
-            }
-        }
-
-        true
-    }
-
-    let generics = tcx.generics_of(assoc_item.def_id);
-    // Chop off any additional substs (RPITIT) substs
-    let substs = &substs[0..generics.count().min(substs.len())];
-    check_substs_compatible_inner(tcx, generics, substs)
 }
 
 fn confirm_impl_trait_in_trait_candidate<'tcx>(

--- a/compiler/rustc_trait_selection/src/traits/util.rs
+++ b/compiler/rustc_trait_selection/src/traits/util.rs
@@ -302,3 +302,44 @@ pub enum TupleArgumentsFlag {
     Yes,
     No,
 }
+
+// Verify that the trait item and its implementation have compatible substs lists
+pub fn check_substs_compatible<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    assoc_item: ty::AssocItem,
+    substs: ty::SubstsRef<'tcx>,
+) -> bool {
+    fn check_substs_compatible_inner<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        generics: &'tcx ty::Generics,
+        args: &'tcx [ty::GenericArg<'tcx>],
+    ) -> bool {
+        if generics.count() != args.len() {
+            return false;
+        }
+
+        let (parent_args, own_args) = args.split_at(generics.parent_count);
+
+        if let Some(parent) = generics.parent
+            && let parent_generics = tcx.generics_of(parent)
+            && !check_substs_compatible_inner(tcx, parent_generics, parent_args) {
+            return false;
+        }
+
+        for (param, arg) in std::iter::zip(&generics.params, own_args) {
+            match (&param.kind, arg.unpack()) {
+                (ty::GenericParamDefKind::Type { .. }, ty::GenericArgKind::Type(_))
+                | (ty::GenericParamDefKind::Lifetime, ty::GenericArgKind::Lifetime(_))
+                | (ty::GenericParamDefKind::Const { .. }, ty::GenericArgKind::Const(_)) => {}
+                _ => return false,
+            }
+        }
+
+        true
+    }
+
+    let generics = tcx.generics_of(assoc_item.def_id);
+    // Chop off any additional substs (RPITIT) substs
+    let substs = &substs[0..generics.count().min(substs.len())];
+    check_substs_compatible_inner(tcx, generics, substs)
+}

--- a/compiler/rustc_ty_utils/src/opaque_types.rs
+++ b/compiler/rustc_ty_utils/src/opaque_types.rs
@@ -6,7 +6,7 @@ use rustc_middle::ty::util::{CheckRegions, NotUniqueParam};
 use rustc_middle::ty::{self, Ty, TyCtxt};
 use rustc_middle::ty::{TypeSuperVisitable, TypeVisitable, TypeVisitor};
 use rustc_span::Span;
-use rustc_type_ir::AliasKind;
+use rustc_trait_selection::traits::check_substs_compatible;
 use std::ops::ControlFlow;
 
 use crate::errors::{DuplicateArg, NotParam};
@@ -36,6 +36,15 @@ impl<'tcx> OpaqueTypeCollector<'tcx> {
         self.tcx.def_span(self.item)
     }
 
+    fn parent_trait_ref(&self) -> Option<ty::TraitRef<'tcx>> {
+        let parent = self.parent()?;
+        if matches!(self.tcx.def_kind(parent), DefKind::Impl { .. }) {
+            Some(self.tcx.impl_trait_ref(parent)?.subst_identity())
+        } else {
+            None
+        }
+    }
+
     fn parent(&self) -> Option<LocalDefId> {
         match self.tcx.def_kind(self.item) {
             DefKind::Fn => None,
@@ -56,7 +65,7 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for OpaqueTypeCollector<'tcx> {
     #[instrument(skip(self), ret, level = "trace")]
     fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<ErrorGuaranteed> {
         match t.kind() {
-            ty::Alias(AliasKind::Opaque, alias_ty) if alias_ty.def_id.is_local() => {
+            ty::Alias(ty::Opaque, alias_ty) if alias_ty.def_id.is_local() => {
                 if !self.seen.insert(alias_ty.def_id.expect_local()) {
                     return ControlFlow::Continue(());
                 }
@@ -98,37 +107,48 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for OpaqueTypeCollector<'tcx> {
                     }
                 }
             }
-            ty::Alias(AliasKind::Projection, alias_ty) => {
-                if let Some(parent) = self.parent() {
-                    trace!(?alias_ty);
-                    let (trait_ref, own_substs) = alias_ty.trait_ref_and_own_substs(self.tcx);
+            ty::Alias(ty::Projection, alias_ty) => {
+                // This avoids having to do normalization of `Self::AssocTy` by only
+                // supporting the case of a method defining opaque types from assoc types
+                // in the same impl block.
+                if let Some(parent_trait_ref) = self.parent_trait_ref() {
+                    // If the trait ref of the associated item and the impl differs,
+                    // then we can't use the impl's identity substitutions below, so
+                    // just skip.
+                    if alias_ty.trait_ref(self.tcx) == parent_trait_ref {
+                        let parent = self.parent().expect("we should have a parent here");
 
-                    trace!(?trait_ref, ?own_substs);
-                    // This avoids having to do normalization of `Self::AssocTy` by only
-                    // supporting the case of a method defining opaque types from assoc types
-                    // in the same impl block.
-                    if trait_ref.self_ty() == self.tcx.type_of(parent).subst_identity() {
-                        for assoc in self.tcx.associated_items(parent).in_definition_order() {
+                        for &assoc in self.tcx.associated_items(parent).in_definition_order() {
                             trace!(?assoc);
-                            if assoc.trait_item_def_id == Some(alias_ty.def_id) {
-                                // We reconstruct the generic args of the associated type within the impl
-                                // from the impl's generics and the generic args passed to the type via the
-                                // projection.
-                                let substs = ty::InternalSubsts::identity_for_item(
-                                    self.tcx,
-                                    parent.to_def_id(),
-                                );
-                                trace!(?substs);
-                                let substs: Vec<_> =
-                                    substs.iter().chain(own_substs.iter().copied()).collect();
-                                trace!(?substs);
-                                // Find opaque types in this associated type.
-                                return self
-                                    .tcx
-                                    .type_of(assoc.def_id)
-                                    .subst(self.tcx, &substs)
-                                    .visit_with(self);
+                            if assoc.trait_item_def_id != Some(alias_ty.def_id) {
+                                continue;
                             }
+
+                            // If the type is further specializable, then the type_of
+                            // is not actually correct below.
+                            if !assoc.defaultness(self.tcx).is_final() {
+                                continue;
+                            }
+
+                            let impl_substs = alias_ty.substs.rebase_onto(
+                                self.tcx,
+                                parent_trait_ref.def_id,
+                                ty::InternalSubsts::identity_for_item(self.tcx, parent),
+                            );
+
+                            if !check_substs_compatible(self.tcx, assoc, impl_substs) {
+                                self.tcx.sess.delay_span_bug(
+                                    self.tcx.def_span(assoc.def_id),
+                                    "item had incorrect substs",
+                                );
+                                return ControlFlow::Continue(());
+                            }
+
+                            return self
+                                .tcx
+                                .type_of(assoc.def_id)
+                                .subst(self.tcx, impl_substs)
+                                .visit_with(self);
                         }
                     }
                 }

--- a/tests/ui/type-alias-impl-trait/impl-trait-in-type-alias-with-bad-substs.rs
+++ b/tests/ui/type-alias-impl-trait/impl-trait-in-type-alias-with-bad-substs.rs
@@ -1,0 +1,28 @@
+#![feature(impl_trait_in_assoc_type)]
+
+// We weren't checking that the trait and impl generics line up in the
+// normalization-shortcut code in `OpaqueTypeCollector`.
+
+use std::ops::Deref;
+
+trait Foo {
+    type Bar<'a>;
+
+    type Baz<'a>;
+
+    fn test<'a>() -> Self::Bar<'a>;
+}
+
+impl Foo for () {
+    type Bar<'a> = impl Deref<Target = Self::Baz<'a>>;
+
+    type Baz<T> = impl Sized;
+    //~^ ERROR type `Baz` has 1 type parameter but its trait declaration has 0 type parameters
+    //~| ERROR unconstrained opaque type
+
+    fn test<'a>() -> Self::Bar<'a> {
+        &()
+    }
+}
+
+fn main() {}

--- a/tests/ui/type-alias-impl-trait/impl-trait-in-type-alias-with-bad-substs.stderr
+++ b/tests/ui/type-alias-impl-trait/impl-trait-in-type-alias-with-bad-substs.stderr
@@ -1,0 +1,20 @@
+error[E0049]: type `Baz` has 1 type parameter but its trait declaration has 0 type parameters
+  --> $DIR/impl-trait-in-type-alias-with-bad-substs.rs:19:14
+   |
+LL |     type Baz<'a>;
+   |              -- expected 0 type parameters
+...
+LL |     type Baz<T> = impl Sized;
+   |              ^ found 1 type parameter
+
+error: unconstrained opaque type
+  --> $DIR/impl-trait-in-type-alias-with-bad-substs.rs:19:19
+   |
+LL |     type Baz<T> = impl Sized;
+   |                   ^^^^^^^^^^
+   |
+   = note: `Baz` must be used in combination with a concrete type within the same impl
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0049`.

--- a/tests/ui/type-alias-impl-trait/not-matching-trait-refs-isnt-defining.rs
+++ b/tests/ui/type-alias-impl-trait/not-matching-trait-refs-isnt-defining.rs
@@ -1,0 +1,33 @@
+#![feature(impl_trait_in_assoc_type)]
+
+trait Foo<T> {
+    type Assoc;
+
+    fn test() -> u32;
+}
+
+struct DefinesOpaque;
+impl Foo<DefinesOpaque> for () {
+    type Assoc = impl Sized;
+
+    // This test's return type is `u32`, *not* the opaque that is defined above.
+    // Previously we were only checking that the self type of the assoc matched,
+    // but this doesn't account for other impls with different trait substs.
+    fn test() -> <() as Foo<NoOpaques>>::Assoc {
+        let _: <Self as Foo<DefinesOpaque>>::Assoc = "";
+        //~^ ERROR mismatched types
+
+        1
+    }
+}
+
+struct NoOpaques;
+impl Foo<NoOpaques> for () {
+    type Assoc = u32;
+
+    fn test() -> u32 {
+        1
+    }
+}
+
+fn main() {}

--- a/tests/ui/type-alias-impl-trait/not-matching-trait-refs-isnt-defining.stderr
+++ b/tests/ui/type-alias-impl-trait/not-matching-trait-refs-isnt-defining.stderr
@@ -1,0 +1,22 @@
+error[E0308]: mismatched types
+  --> $DIR/not-matching-trait-refs-isnt-defining.rs:17:54
+   |
+LL |     type Assoc = impl Sized;
+   |                  ---------- the expected opaque type
+...
+LL |         let _: <Self as Foo<DefinesOpaque>>::Assoc = "";
+   |                -----------------------------------   ^^ expected opaque type, found `&str`
+   |                |
+   |                expected due to this
+   |
+   = note: expected opaque type `<() as Foo<DefinesOpaque>>::Assoc`
+                found reference `&'static str`
+note: this item must have the opaque type in its signature in order to be able to register hidden types
+  --> $DIR/not-matching-trait-refs-isnt-defining.rs:16:5
+   |
+LL |     fn test() -> <() as Foo<NoOpaques>>::Assoc {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

